### PR TITLE
modified work_queue_factory to integrate runos into the factory. 

### PR DESCRIPTION
**Pull request by Nekel-Seyew**
Not super portable right now. This pull request is more for demonstrating an idea, to be discussed at HEP meetings.

### CODE REVIEW

File: batch_job/src/work_queue_factory.c [hunk 1]
```diff
@@ -373,6 +375,14 @@ static int submit_worker( struct batch_queue *queue )
 		free(files);
 		files = newfiles;
 	}
+	
+	if(os){
+		char* temp = string_format("python /afs/crc.nd.edu/group/ccl/software/runos/runos.py %s %s",os,cmd);
```
Comment (btovar) [inline batch_job/src/work_queue_factory.c]: I think you want to have runos in your PATH, copy it to scratch_dir, and send it as we send the worker. (See the line that has "which work_queue_worker".

File: batch_job/src/work_queue_factory.c [hunk 2]
```diff
@@ -360,7 +362,7 @@ static int submit_worker( struct batch_queue *queue )
 		cmd = newcmd;
 	}
 
-	char *files = string_format("work_queue_worker");
+	char *files = os ? string_format("") : string_format("work_queue_worker");
```
Comment (btovar) [inline batch_job/src/work_queue_factory.c]: I would make it:

files = os ? string_format("runos.py") : string_format("work_queue_worker");

Ideally, the factory should not send anything that depends on a global path.

File: batch_job/src/work_queue_factory.c [hunk 3]
```diff
@@ -322,7 +324,7 @@ static void set_worker_resources_options( struct batch_queue *queue )
 static int submit_worker( struct batch_queue *queue )
 {
 	char *cmd;
-	const char *worker = "./work_queue_worker";
+	const char *worker = os ? "work_queue_worker": "./work_queue_worker";
```
Comment (btovar) [inline batch_job/src/work_queue_factory.c]: This assumes the worker is always in the PATH in os. Is this always true?

File: batch_job/src/work_queue_factory.c [hunk 4]
```diff
@@ -100,6 +100,8 @@ static char *wrapper_command = 0;
 static char *wrapper_input = 0;
 static char *worker_command = 0;
 
+static char *os = NULL;
```
Comment (btovar) [inline batch_job/src/work_queue_factory.c]: I suggest to rename this variable to something more explicit, e.g., runos_opsys.

File: batch_job/src/work_queue_factory.c (hunk 2 above)
Comment (btovar) [reply batch_job/src/work_queue_factory.c]: That is, it should work as if we passed --wrapper-file runos.py to the factory.

File: batch_job/src/work_queue_factory.c (hunk 3 above)
Comment (Nekel-Seyew) [reply batch_job/src/work_queue_factory.c]: if it's in the case of being inside of singularity (as all cctools are in the path of my custom images), or now as being inside the wrapper of vc3-builder, then yes.

**Comment by btovar** [comment]
This seems very specific to ND.

I can deploy it so that ND CMS can use in the meanwhile, but I don't think it should be merged in its current form to master.

File: batch_job/src/work_queue_factory.c [hunk 5]
```diff
@@ -1161,6 +1187,13 @@ int main(int argc, char *argv[])
 		}
 	}
 
+	memset(cmd,'\0',1024*sizeof(char));
```
Comment (btovar) [inline batch_job/src/work_queue_factory.c]: bzero(cmd, sizeof(*cmd));

Comment (Nekel-Seyew) [reply batch_job/src/work_queue_factory.c]: Isn't bzero a gnu specific function in string.h?

File: batch_job/src/work_queue_factory.c [hunk 6]
```diff
@@ -98,6 +98,8 @@ static char *wrapper_command = 0;
 static char *wrapper_input = 0;
 static char *worker_command = 0;
 
+static char *os = NULL;
```
Comment (btovar) [inline batch_job/src/work_queue_factory.c]: change to *runos_os = 0;

File: batch_job/src/work_queue_factory.c [hunk 7]
```diff
@@ -329,7 +331,7 @@ static void set_worker_resources_options( struct batch_queue *queue )
 static int submit_worker( struct batch_queue *queue )
 {
 	char *cmd;
-	const char *worker = "./work_queue_worker";
+	const char *worker = "work_queue_worker";
```
Comment (btovar) [inline batch_job/src/work_queue_factory.c]: This change breaks work_queue_factory when not using runos.

File: batch_job/src/work_queue_factory.c [hunk 8]
```diff
@@ -365,7 +367,7 @@ static int submit_worker( struct batch_queue *queue )
 		cmd = newcmd;
 	}
 
-	char *files = string_format("work_queue_worker");
+	char *files=string_format("%s","");// = os ? string_format("%s","") : string_format("work_queue_worker");
```
Comment (btovar) [inline batch_job/src/work_queue_factory.c]: instead use: xxstrudp("");

File: batch_job/src/work_queue_factory.c [hunk 9]
```diff
@@ -378,6 +380,23 @@ static int submit_worker( struct batch_queue *queue )
 		free(files);
 		files = newfiles;
 	}
+	
+	if(os){
```
Comment (btovar) [inline batch_job/src/work_queue_factory.c]: if(runos_os) {

File: batch_job/src/work_queue_factory.c [hunk 10]
```diff
@@ -378,6 +380,23 @@ static int submit_worker( struct batch_queue *queue )
...
+		char* vc3_cmd = string_format("./vc3-builder --require cctools-statics -- %s",cmd);
+		char* temp = string_format("python /afs/crc.nd.edu/group/ccl/software/runos/runos.py %s %s",os,vc3_cmd);
+		free(vc3_cmd);
+		free(cmd);
+		cmd = temp;
+		temp = string_format("%s,%s",files,"vc3-builder");
+		free(files);
+		files = temp;
+	}else{
+		char* vc3_cmd = string_format("./vc3-builder --require cctools-statics -- %s",cmd);
```
Comment (btovar) [inline batch_job/src/work_queue_factory.c]: If we are not using runos, we do not want to use the builder.

File: batch_job/src/work_queue_factory.c (hunk 5 above)
Comment (btovar) [reply batch_job/src/work_queue_factory.c]: Yes, I did not know it was deprecated... I liked it had less confusing order of arguments.

Do remove the 1024. Instead use sizeof(cmd);

File: batch_job/src/work_queue_factory.c [hunk 11]
```diff
@@ -1263,6 +1288,14 @@ int main(int argc, char *argv[])
 		}
 	}
 
+	memset(cmd,'\0',1024*sizeof(char));
+	sprintf(cmd,"cp \"$(which vc3-builder)\" '%s'",scratch_dir);
+	int k = system(cmd);
+	if(k){
+		fprintf(stderr,"can't copy vc3-builder!: %i\n",k);
```
Comment (btovar) [inline batch_job/src/work_queue_factory.c]: If so, say that the builder should be in the PATH.

File: batch_job/src/work_queue_factory.c [hunk 12]
```diff
@@ -381,21 +381,19 @@ static int submit_worker( struct batch_queue *queue )
...
 		char* vc3_cmd = string_format("./vc3-builder --require cctools-statics -- %s",cmd);
-		char* temp = string_format("python /afs/crc.nd.edu/group/ccl/software/runos/runos.py %s %s",os,vc3_cmd);
+		char* temp = string_format("python /afs/crc.nd.edu/group/ccl/software/runos/runos.py %s %s",runos_os,vc3_cmd);
 		free(vc3_cmd);
 		free(cmd);
 		cmd = temp;
 		temp = string_format("%s,%s",files,"vc3-builder");
 		free(files);
 		files = temp;
 	}else{
```
Comment (btovar) [inline batch_job/src/work_queue_factory.c]: Where is cmd defined for the regular case?

File: batch_job/src/work_queue_factory.c [hunk 13]
```diff
@@ -381,21 +381,19 @@ static int submit_worker( struct batch_queue *queue )
 		files = newfiles;
 	}
 	
-	if(os){
+	if(runos_os){
 		char* vc3_cmd = string_format("./vc3-builder --require cctools-statics -- %s",cmd);
-		char* temp = string_format("python /afs/crc.nd.edu/group/ccl/software/runos/runos.py %s %s",os,vc3_cmd);
+		char* temp = string_format("python /afs/crc.nd.edu/group/ccl/software/runos/runos.py %s %s",runos_os,vc3_cmd);
```
Comment (btovar) [inline batch_job/src/work_queue_factory.c]: make a #define CCTOOLS_RUNOS_PATH /afs/... at the top of this file, and then use "python " CCTOOLS_RUNOS_PATH " %s %s", etc

File: batch_job/src/work_queue_factory.c (hunk 12 above)
Comment (Nekel-Seyew) [reply batch_job/src/work_queue_factory.c]: cmd is only ever wrapped around if it's changed. So the root cmd is defined first by "if catalog else" statements, and then only modified from there. The else statement here for the runos_os if statement simply verifies that we're not using runos or vc3-builder, so to use the normal work_queue_worker.

File: batch_job/src/work_queue_factory.c [hunk 14]
```diff
@@ -1248,21 +1271,30 @@ int main(int argc, char *argv[])
...
-		sprintf(cmd,"cp \"$(which work_queue_worker)\" '%s'",scratch_dir);
+		cmd = string_format("cp \"$(which work_queue_worker)\" '%s'",scratch_dir);
 		if (system(cmd)) {
 			fprintf(stderr, "work_queue_factory: please add work_queue_worker to your PATH.\n");
 			exit(EXIT_FAILURE);
 		}
 	}
 
+	if(runos_os) {
+		cmd = string_format("cp \"$(which vc3-builder)\" '%s'", scratch_dir);
```
Comment (btovar) [inline batch_job/src/work_queue_factory.c]: free(cmd), since it was allocated in the if-else block above.

Comment (Nekel-Seyew) [reply batch_job/src/work_queue_factory.c]: Done! Also changed it in other places to reflect that change.

**Comment by dthain** [comment]
@btovar Merge if you are satisfied.

**Pull request merged**